File: te_greens.py
"""
TE Green's function engine for multilayer structures.

Goal:
    Compute G_yy(l_obs, l_src, z_obs, z_src, k_parallel)
    using a clean 5-level architecture:

        Level 1: q list (per layer)
        Level 2: TE reflection recursion (R_down, R_up)
        Level 3: same-layer f1, f2 amplitudes (local)
        Level 4: propagation between layers (down/up)
        Level 5: assemble full TE Green's function G_yy

This is a collaborative development skeleton.
"""


import numpy as np
from multilayer_reflectance import RF_multilayer, RB_multilayer  

# ------------------------------------------------------------
# Level 1: q(k_parallel) per layer
# ------------------------------------------------------------
def compute_q_list(n_list, k0, kp):
    """  Return q_list for all layers."""
    eps_list = np.array(n_list)**2
    q_list = np.sqrt(kp**2 - eps_list * k0**2 + 0j) / k0
    return q_list

def compute_interface_params_up_down(n_list, d_list, k0, kp, polarization="TE"):
    """
    Returns two lists:
        params_down[L] : for crossing interface L downward (layer L → L+1)
        params_up[L]   : for crossing interface L upward   (layer L+1 → L)

    Each params_down[L] / params_up[L] contains:
        q_up      : q in layer L      (source side)
        q_down    : q in layer L+1    (destination side)
        x         : exp(-q_up * k0 * d_list[L])  # SAME for up/down
        wp, wq    : jump coefficients for recursion
    """
    N = len(n_list)
    eps = np.array(n_list) ** 2
    kp2 = kp ** 2
    params_down = []
    params_up = []

    for L in range(N - 1):
        q_up_layer   = np.sqrt(kp2 - eps[L]   * k0**2 + 0j) / k0
        q_down_layer = np.sqrt(kp2 - eps[L+1] * k0**2 + 0j) / k0
        xL = np.exp(- q_up_layer * k0 * d_list[L])

        if polarization.upper() == "TE":
            wn_down = q_down_layer / q_up_layer
            wn_up   = q_up_layer   / q_down_layer
        else:   # TM
            wn_down = (q_down_layer / q_up_layer) * (eps[L] / eps[L+1])
            wn_up   = (q_up_layer   / q_down_layer) * (eps[L+1] / eps[L])

        wm = 1.0
        wp_down = wm + wn_down
        wq_down = wm - wn_down
        wp_up = wm + wn_up
        wq_up = wm - wn_up

        params_down.append({"q_up": q_up_layer,"q_down": q_down_layer,"x": xL,"wp": wp_down,"wq": wq_down})
        params_up.append({"q_up": q_up_layer,"q_down": q_down_layer,"x": xL,"wp": wp_up,"wq": wq_up})

    return params_down, params_up

def compute_RF_all(n_list, d_list, k0, kp, polarization="TE"):
    """
    RF_all[j] = reflection seen when standing at TOP of layer j, looking downward.
    Implemented using params_down (wp, wq, x), equivalent to original RF_multilayer().
    """
    params_down, params_up = compute_interface_params_up_down(n_list, d_list, k0, kp, polarization)
    N = len(n_list)
    RF_all = [0.0 + 0.0j] * N
    r_eff = 0.0 + 0.0j

    # recursion from bottom to top
    # interfaces are L = 0...(N-2), between layer L and L+1
    for L in range(N-2, -1, -1):
        p = params_down[L]
        wp = p["wp"]
        wq = p["wq"]
        x  = p["x"]      # exponential through layer L (downward)

        # downward recursion formula:
        # r_new = x^2 * (wq + wp * r_old) / (wp + wq * r_old)
        num = wq + wp * r_eff
        den = wp + wq * r_eff
        r_eff = x * (num / den) * x   # x^2 * (bl/al)
        RF_all[L] = r_eff

    RF_all[N-1] = 0.0 + 0.0j   # bottom semi-infinite region
    return RF_all


def compute_RB_all(n_list, d_list, k0, kp, polarization="TE"):
    """
    RB_all[j] = reflection seen when standing at BOTTOM of layer j, looking upward.
    Implemented using params_up (wp, wq, x), equivalent to original RB_multilayer().
    """
    params_down, params_up = compute_interface_params_up_down(n_list, d_list, k0, kp, polarization)
    N = len(n_list)
    RB_all = [0.0 + 0.0j] * N
    r_eff = 0.0 + 0.0j

    # recursion from top to bottom
    for L in range(0, N-1):
        p = params_up[L]
        wp = p["wp"]
        wq = p["wq"]
        x  = p["x"]      # exponential through layer L (upward)
        xr = x * r_eff * x
        num = wq + wp * xr
        den = wp + wq * xr
        r_eff = num / den
        RB_all[L+1] = r_eff

    RB_all[0] = 0.0 + 0.0j   # top semi-infinite region
    return RB_all

# ------------------------------------------------------------
# Level 3: same-layer source amplitudes f1, f2 (local)
# ------------------------------------------------------------
def TE_f1yf2y_same_layer(q_list, R_down, R_up, layer_src, z_src, k0):
    """
    Compute TE source amplitudes (f1y, f2y) inside the source layer.
    f1y : downward-propagating TE wave amplitude
    f2y : upward-propagating TE wave amplitude
    """
    q  = q_list[layer_src]
    RF = R_down[layer_src]
    RB = R_up[layer_src]
    ez  = np.exp(+q * k0 * z_src)
    emz = np.exp(-q * k0 * z_src)
    den = 2 * q * k0 * (1 - RF * RB)
    f1y = ( ez  + emz * RB ) / den
    f2y = ( emz + ez  * RF ) / den
    return f1y, f2y


# ------------------------------------------------------------
# Level 4: propagate to another layer
# ------------------------------------------------------------
def propagate_down_TE(f1_src, q_list, R_down, R_up, layer_src, layer_obs, k0, d_list):
    if layer_obs <= layer_src:
        raise ValueError("propagate_down_TE requires layer_obs > layer_src")

    f = f1_src
    for l in range(layer_src, layer_obs):
        d_l = d_list[l] if l < len(d_list) else 0.0
        ql  = q_list[l]
        qlp = q_list[l+1]
        wn = qlp / ql
        wp = 1.0 + wn          # M_l^+
        wq = 1.0 - wn          # M_l^-
        Rnext = R_down[l+1]    # R_{l+1}
        T = 1.0 / (wp + wq * Rnext)       # T_l (paper)
        x = np.exp(-ql * k0 * d_l)        # e^{-q_l d_l}
        f = T * x * f                     # f_{l+1} = T_l e^{-q_l d_l} f_l
    return f

def propagate_up_TE(f2_src, q_list, R_up, layer_src, layer_obs, k0, d_list):  
    if layer_obs >= layer_src:
        raise ValueError("propagate_up_TE requires layer_obs < layer_src")
    
    f = f2_src
    for l in range(layer_src - 1, layer_obs - 1, -1):
        d_l = d_list[l] if l < len(d_list) else 0.0
        ql   = q_list[l]
        qlp1 = q_list[l + 1]
        x = np.exp(-ql * k0 * d_l)
        wn = ql / qlp1
        mp = 1.0 + wn     # m^+
        mm = 1.0 - wn     # m^-
        r_l = R_up[l]
        T = 1.0 / (mp + mm * x * r_l * x)
        f = x * T * f
    return f


# ------------------------------------------------------------
# Level 5: assemble G_yy
# ------------------------------------------------------------
def gyy_TE(n_list, d_list, layer_src, z_src, layer_obs, z_obs, k0, kp):

    q_list = compute_q_list(n_list, k0, kp)
    R_down = compute_RF_all(n_list, d_list, k0, kp, polarization="TE")
    R_up   = compute_RB_all(n_list, d_list, k0, kp, polarization="TE")
    f1y_src, f2y_src = TE_f1yf2y_same_layer(q_list, R_down, R_up, layer_src, z_src, k0)
    q_obs  = q_list[layer_obs]
    RF_obs = R_down[layer_obs]
    RB_obs = R_up[layer_obs]
    Gyy = 0.0 + 0.0j

    if layer_obs == layer_src:
        q  = q_list[layer_src]
        RF = R_down[layer_src]
        RB = R_up[layer_src]
        direct = (1.0 / (2.0 * q * k0)) * np.exp(-q * k0 * np.abs(z_obs - z_src))
        cavity = ( np.exp(+q * k0 * z_obs) * RF * f1y_src + np.exp(-q * k0 * z_obs) * RB * f2y_src )
        Gyy = direct + cavity

    elif layer_obs > layer_src:
        f1y_at_obs_top = propagate_down_TE( f1y_src, q_list, R_down,layer_src, layer_obs, k0, d_list )
        Gyy = ( np.exp(-q_obs * k0 * z_obs) + np.exp(+q_obs * k0 * z_obs) * RF_obs) * f1y_at_obs_top

    else:
        f2y_at_obs_top = propagate_up_TE( f2y_src, q_list, R_up,layer_src, layer_obs, k0, d_list )
        Gyy = (np.exp(+q_obs * k0 * z_obs)+ np.exp(-q_obs * k0 * z_obs) * RB_obs) * f2y_at_obs_top

    return Gyy

def dgyy_dzobs_TE(n_list, d_list,layer_src, z_src,layer_obs, z_obs,k0, kp,tol: float = 1e-12,):
    """
    d/dz_obs of Gyy for TE polarization.

    This version raises an error for the same-layer cusp at z_obs == z_src
    (or |z_obs - z_src| < tol), which is appropriate for Poynting-flux
    surface integrals (observation surface should not pass through the source).
    """

    q_list = compute_q_list(n_list, k0, kp)
    R_down = compute_RF_all(n_list, d_list, k0, kp, polarization="TE")
    R_up   = compute_RB_all(n_list, d_list, k0, kp, polarization="TE")

    f1y_src, f2y_src = TE_f1yf2y_same_layer(q_list, R_down, R_up, layer_src, z_src, k0)

    q_obs  = q_list[layer_obs]
    RF_obs = R_down[layer_obs]
    RB_obs = R_up[layer_obs]

    # -------------------------
    # Case A: same layer
    # -------------------------
    if layer_obs == layer_src:
        q  = q_list[layer_src]
        RF = R_down[layer_src]
        RB = R_up[layer_src]

        a  = q * k0
        dz = (z_obs - z_src)

        if np.abs(dz) < tol:
            raise ValueError(
                "dgyy_dzobs_TE (same-layer): |z_obs - z_src| < tol -> cusp at the source.\n"
                "Move the observation surface away from the source location."
            )

        s = np.sign(dz)  # ±1 for real dz

        # direct = (1/(2 q k0)) * exp(-a*|dz|)
        # ddirect = (1/(2 q k0)) * exp(-a|dz|) * (-a*sign(dz)) = -(1/2)*exp(-a|dz|)*sign(dz)
        ddirect = (-0.5) * np.exp(-a * np.abs(dz)) * s

        # cavity = exp(+a z_obs)*RF*f1y_src + exp(-a z_obs)*RB*f2y_src
        # dcavity = a*( exp(+a z_obs)*RF*f1y_src - exp(-a z_obs)*RB*f2y_src )
        dcavity = a * (
            np.exp(+a * z_obs) * RF * f1y_src
            - np.exp(-a * z_obs) * RB * f2y_src
        )

        return ddirect + dcavity

    # -------------------------
    # Case B: obs below src
    # -------------------------
    if layer_obs > layer_src:
        f1y_at_obs_top = propagate_down_TE(
            f1y_src, q_list, R_down,
            layer_src, layer_obs,
            k0, d_list
        )

        a_obs = q_obs * k0
        # dress = exp(-a_obs z_obs) + exp(+a_obs z_obs)*RF_obs
        # ddress = -a_obs exp(-a_obs z_obs) + a_obs exp(+a_obs z_obs)*RF_obs
        ddress = (-a_obs) * np.exp(-a_obs * z_obs) + (a_obs) * np.exp(+a_obs * z_obs) * RF_obs

        return ddress * f1y_at_obs_top

    # -------------------------
    # Case C: obs above src
    # -------------------------
    f2y_at_obs_top = propagate_up_TE(
        f2y_src, q_list, R_up,
        layer_src, layer_obs,
        k0, d_list
    )

    a_obs = q_obs * k0
    # dress = exp(+a_obs z_obs) + exp(-a_obs z_obs)*RB_obs
    # ddress = a_obs exp(+a_obs z_obs) - a_obs exp(-a_obs z_obs)*RB_obs
    ddress = (a_obs) * np.exp(+a_obs * z_obs) + (-a_obs) * np.exp(-a_obs * z_obs) * RB_obs

    return ddress * f2y_at_obs_top


if __name__ == "__main__":
    import matplotlib.pyplot as plt

    n_list = [1.0, 1.5, 1.3, 1.0]  # 四層結構
    d_list = [0.0, 2000, 1500]   # 四層結構
    wls = np.linspace(500, 900, 200)
    diff_RF = []
    diff_RB = []

    for pol in ["TE", "TM"]:
        diff_RF = []
        diff_RB = []

        for wl in wls:
            k0 = 2*np.pi / wl
            kp = 0.3*k0

            # OLD version
            RF_old = [RF_multilayer(n_list, d_list, pol, k0, kp, j)
                      for j in range(len(n_list))]
            RB_old = [RB_multilayer(n_list, d_list, pol, k0, kp, j)
                      for j in range(len(n_list))]

            # NEW version
            RF_new = compute_RF_all(n_list, d_list, k0, kp, pol)
            RB_new = compute_RB_all(n_list, d_list, k0, kp, pol)

            diff_RF.append(np.max(np.abs(np.array(RF_old) - np.array(RF_new))))
            diff_RB.append(np.max(np.abs(np.array(RB_old) - np.array(RB_new))))

        print(f"=== polarization: {pol} ===")
        print("max |RF_old - RF_new| =", np.max(diff_RF))
        print("max |RB_old - RB_new| =", np.max(diff_RB))
        print()

File: fortran/te_greens.f90
! te_greens.f90
! TE Green's function engine for multilayer structures

module te_greens
    use multilayer_reflectance
    implicit none
    
contains

    ! Level 1: compute q list for all layers
    subroutine compute_q_list(n_list, N, k0, kp, q_list)
        integer, intent(in) :: N
        real(dp), intent(in) :: n_list(N), k0, kp
        complex(dp), intent(out) :: q_list(N)
        real(dp) :: eps_list(N)
        integer :: i
        
        eps_list = n_list**2
        do i = 1, N
            q_list(i) = sqrt(cmplx(kp**2 - eps_list(i) * k0**2, 0.0_dp, kind=dp)) / k0
        end do
    end subroutine compute_q_list

    ! Compute RF for all layers
    subroutine compute_RF_all(n_list, d_list, N, k0, kp, polarization, RF_all)
        integer, intent(in) :: N
        real(dp), intent(in) :: n_list(N), d_list(N-1), k0, kp
        character(len=*), intent(in) :: polarization
        complex(dp), intent(out) :: RF_all(N)
        
        real(dp) :: eps(N), kp2
        complex(dp) :: q_up, q_down, wn, wp, wq, x_val, r_eff, num, den
        integer :: L
        
        eps = n_list**2
        kp2 = kp**2
        r_eff = cmplx(0.0_dp, 0.0_dp, kind=dp)
        
        ! Recursion from bottom to top
        do L = N-1, 1, -1
            q_up = sqrt(cmplx(kp2 - eps(L) * k0**2, 0.0_dp, kind=dp)) / k0
            q_down = sqrt(cmplx(kp2 - eps(L+1) * k0**2, 0.0_dp, kind=dp)) / k0
            x_val = exp(-q_up * k0 * d_list(L))
            
            if (polarization == "TE" .or. polarization == "te") then
                wn = q_down / q_up
            else  ! TM
                wn = (q_down / q_up) * (eps(L) / eps(L+1))
            end if
            
            wp = 1.0_dp + wn
            wq = 1.0_dp - wn
            
            num = wq + wp * r_eff
            den = wp + wq * r_eff
            r_eff = x_val * (num / den) * x_val
            RF_all(L) = r_eff
        end do
        
        RF_all(N) = cmplx(0.0_dp, 0.0_dp, kind=dp)
    end subroutine compute_RF_all

    ! Compute RB for all layers
    subroutine compute_RB_all(n_list, d_list, N, k0, kp, polarization, RB_all)
        integer, intent(in) :: N
        real(dp), intent(in) :: n_list(N), d_list(N-1), k0, kp
        character(len=*), intent(in) :: polarization
        complex(dp), intent(out) :: RB_all(N)
        
        real(dp) :: eps(N), kp2
        complex(dp) :: q_up, q_down, wn, wp, wq, x_val, r_eff, xr, num, den
        integer :: L
        
        eps = n_list**2
        kp2 = kp**2
        r_eff = cmplx(0.0_dp, 0.0_dp, kind=dp)
        
        RB_all(1) = cmplx(0.0_dp, 0.0_dp, kind=dp)
        
        ! Recursion from top to bottom
        do L = 1, N-1
            q_up = sqrt(cmplx(kp2 - eps(L) * k0**2, 0.0_dp, kind=dp)) / k0
            q_down = sqrt(cmplx(kp2 - eps(L+1) * k0**2, 0.0_dp, kind=dp)) / k0
            x_val = exp(-q_up * k0 * d_list(L))
            
            if (polarization == "TE" .or. polarization == "te") then
                wn = q_up / q_down
            else  ! TM
                wn = (q_up / q_down) * (eps(L+1) / eps(L))
            end if
            
            wp = 1.0_dp + wn
            wq = 1.0_dp - wn
            
            xr = x_val * r_eff * x_val
            num = wq + wp * xr
            den = wp + wq * xr
            r_eff = num / den
            RB_all(L+1) = r_eff
        end do
    end subroutine compute_RB_all

    ! Level 3: same-layer source amplitudes
    subroutine TE_f1yf2y_same_layer(q_list, R_down, R_up, N, layer_src, z_src, k0, f1y, f2y)
        integer, intent(in) :: N, layer_src
        complex(dp), intent(in) :: q_list(N), R_down(N), R_up(N)
        real(dp), intent(in) :: z_src, k0
        complex(dp), intent(out) :: f1y, f2y
        
        complex(dp) :: q, RF, RB, ez, emz, den
        
        q = q_list(layer_src)
        RF = R_down(layer_src)
        RB = R_up(layer_src)
        
        ez = exp(q * k0 * z_src)
        emz = exp(-q * k0 * z_src)
        den = 2.0_dp * q * k0 * (1.0_dp - RF * RB)
        
        f1y = (ez + emz * RB) / den
        f2y = (emz + ez * RF) / den
    end subroutine TE_f1yf2y_same_layer

    ! Level 4: propagate downward
    function propagate_down_TE(f1_src, q_list, R_down, N, layer_src, layer_obs, k0, d_list) result(f)
        integer, intent(in) :: N, layer_src, layer_obs
        complex(dp), intent(in) :: f1_src, q_list(N), R_down(N)
        real(dp), intent(in) :: k0, d_list(N-1)
        complex(dp) :: f
        
        complex(dp) :: ql, qlp, wn, wp, wq, Rnext, T, x_val
        real(dp) :: d_l
        integer :: l
        
        f = f1_src
        
        do l = layer_src, layer_obs-1
            d_l = d_list(l)
            ql = q_list(l)
            qlp = q_list(l+1)
            
            wn = qlp / ql
            wp = 1.0_dp + wn
            wq = 1.0_dp - wn
            Rnext = R_down(l+1)
            
            T = 1.0_dp / (wp + wq * Rnext)
            x_val = exp(-ql * k0 * d_l)
            f = T * x_val * f
        end do
    end function propagate_down_TE

    ! Level 4: propagate upward
    function propagate_up_TE(f2_src, q_list, R_up, N, layer_src, layer_obs, k0, d_list) result(f)
        integer, intent(in) :: N, layer_src, layer_obs
        complex(dp), intent(in) :: f2_src, q_list(N), R_up(N)
        real(dp), intent(in) :: k0, d_list(N-1)
        complex(dp) :: f
        
        complex(dp) :: ql, qlp1, wn, mp, mm, r_l, T, x_val
        real(dp) :: d_l
        integer :: l
        
        f = f2_src
        
        do l = layer_src-1, layer_obs, -1
            d_l = d_list(l)
            ql = q_list(l)
            qlp1 = q_list(l+1)
            
            x_val = exp(-ql * k0 * d_l)
            wn = ql / qlp1
            mp = 1.0_dp + wn
            mm = 1.0_dp - wn
            r_l = R_up(l)
            
            T = 1.0_dp / (mp + mm * x_val * r_l * x_val)
            f = x_val * T * f
        end do
    end function propagate_up_TE

    ! Level 5: assemble G_yy
    function gyy_TE(n_list, d_list, N, layer_src, z_src, layer_obs, z_obs, k0, kp) result(Gyy)
        integer, intent(in) :: N, layer_src, layer_obs
        real(dp), intent(in) :: n_list(N), d_list(N-1), z_src, z_obs, k0, kp
        complex(dp) :: Gyy
        
        complex(dp) :: q_list(N), R_down(N), R_up(N)
        complex(dp) :: f1y_src, f2y_src, q_obs, RF_obs, RB_obs
        complex(dp) :: q, RF, RB, direct, cavity, f1y_at_obs, f2y_at_obs
        
        call compute_q_list(n_list, N, k0, kp, q_list)
        call compute_RF_all(n_list, d_list, N, k0, kp, "TE", R_down)
        call compute_RB_all(n_list, d_list, N, k0, kp, "TE", R_up)
        
        call TE_f1yf2y_same_layer(q_list, R_down, R_up, N, layer_src, z_src, k0, f1y_src, f2y_src)
        
        q_obs = q_list(layer_obs)
        RF_obs = R_down(layer_obs)
        RB_obs = R_up(layer_obs)
        
        if (layer_obs == layer_src) then
            q = q_list(layer_src)
            RF = R_down(layer_src)
            RB = R_up(layer_src)
            
            direct = (1.0_dp / (2.0_dp * q * k0)) * exp(-q * k0 * abs(z_obs - z_src))
            cavity = exp(q * k0 * z_obs) * RF * f1y_src + exp(-q * k0 * z_obs) * RB * f2y_src
            Gyy = direct + cavity
            
        else if (layer_obs > layer_src) then
            f1y_at_obs = propagate_down_TE(f1y_src, q_list, R_down, N, layer_src, layer_obs, k0, d_list)
            Gyy = (exp(-q_obs * k0 * z_obs) + exp(q_obs * k0 * z_obs) * RF_obs) * f1y_at_obs
            
        else
            f2y_at_obs = propagate_up_TE(f2y_src, q_list, R_up, N, layer_src, layer_obs, k0, d_list)
            Gyy = (exp(q_obs * k0 * z_obs) + exp(-q_obs * k0 * z_obs) * RB_obs) * f2y_at_obs
        end if
    end function gyy_TE

end module te_greens

! Simple test program
program test_te_greens
    use te_greens
    implicit none
    
    integer, parameter :: N = 4
    real(dp) :: n_list(N), d_list(N-1)
    real(dp) :: z_src, z_obs, k0, kp, wl
    complex(dp) :: Gyy
    
    ! Setup
    n_list = [1.0_dp, 1.5_dp, 1.3_dp, 1.0_dp]
    d_list = [0.0_dp, 2000.0_dp, 1500.0_dp]
    
    z_src = 100.0_dp
    z_obs = 200.0_dp
    wl = 650.0_dp
    k0 = 2.0_dp * 3.14159265359_dp / wl
    kp = 0.3_dp * k0
    
    Gyy = gyy_TE(n_list, d_list, N, 1, z_src, 1, z_obs, k0, kp)
    
    print *, 'Gyy = ', real(Gyy), aimag(Gyy)
    
end program test_te_greens
